In [1]:
%autosave 0

Autosave disabled


# TV Script Generation

In this project, you'll generate your own [Seinfeld](https://en.wikipedia.org/wiki/Seinfeld) TV scripts using RNNs.  You'll be using part of the [Seinfeld dataset](https://www.kaggle.com/thec03u5/seinfeld-chronicles#scripts.csv) of scripts from 9 seasons.  The Neural Network you'll build will generate a new ,"fake" TV script, based on patterns it recognizes in this training data.

## Get the Data

The data is already provided for you in `./data/Seinfeld_Scripts.txt` and you're encouraged to open that file and look at the text. 
>* As a first step, we'll load in this data and look at some samples. 
* Then, you'll be tasked with defining and training an RNN to generate a new script!

In [2]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# load in data
import helper
data_dir = './data/Seinfeld_Scripts.txt'
text = helper.load_data(data_dir)

## Explore the Data
Play around with `view_line_range` to view different parts of the data. This will give you a sense of the data you'll be working with. You can see, for example, that it is all lowercase text, and each new line of dialogue is separated by a newline character `\n`.

In [3]:
view_line_range = (0, 10)

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
import numpy as np

print('Dataset Stats')
print('Roughly the number of unique words: {}'.format(len({word: None for word in text.split()})))

lines = text.split('\n')
print('Number of lines: {}'.format(len(lines)))
word_count_line = [len(line.split()) for line in lines]
print('Average number of words in each line: {}'.format(np.average(word_count_line)))

print()
print('The lines {} to {}:'.format(*view_line_range))
print('\n'.join(text.split('\n')[view_line_range[0]:view_line_range[1]]))

Dataset Stats
Roughly the number of unique words: 46367
Number of lines: 109233
Average number of words in each line: 5.544240293684143

The lines 0 to 10:
jerry: do you know what this is all about? do you know, why were here? to be out, this is out...and out is one of the single most enjoyable experiences of life. people...did you ever hear people talking about we should go out? this is what theyre talking about...this whole thing, were all out now, no one is home. not one person here is home, were all out! there are people trying to find us, they dont know where we are. (on an imaginary phone) did you ring?, i cant find him. where did he go? he didnt tell me where he was going. he must have gone out. you wanna go out you get ready, you pick out the clothes, right? you take the shower, you get all ready, get the cash, get your friends, the car, the spot, the reservation...then youre standing around, what do you do? you go we gotta be getting back. once youre out, you wanna get back! y

---
## Implement Pre-processing Functions
The first thing to do to any dataset is pre-processing.  Implement the following pre-processing functions below:
- Lookup Table
- Tokenize Punctuation

### Lookup Table
To create a word embedding, you first need to transform the words to ids.  In this function, create two dictionaries:
- Dictionary to go from the words to an id, we'll call `vocab_to_int`
- Dictionary to go from the id to word, we'll call `int_to_vocab`

Return these dictionaries in the following **tuple** `(vocab_to_int, int_to_vocab)`

In [4]:
import problem_unittests as tests
from collections import Counter

def create_lookup_tables(text):
    """
    Create lookup tables for vocabulary
    :param text: The text of tv scripts split into words
    :return: A tuple of dicts (vocab_to_int, int_to_vocab)
    """

    vocab = set(text)

    int_to_vocab = {}
    vocab_to_int = {}

    for idx, word in enumerate(vocab):
        int_to_vocab[idx] = word
        vocab_to_int[word] = idx

    return (vocab_to_int, int_to_vocab)

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_create_lookup_tables(create_lookup_tables)

Tests Passed


### Tokenize Punctuation
We'll be splitting the script into a word array using spaces as delimiters.  However, punctuations like periods and exclamation marks can create multiple ids for the same word. For example, "bye" and "bye!" would generate two different word ids.

Implement the function `token_lookup` to return a dict that will be used to tokenize symbols like "!" into "||Exclamation_Mark||".  Create a dictionary for the following symbols where the symbol is the key and value is the token:
- Period ( **.** )
- Comma ( **,** )
- Quotation Mark ( **"** )
- Semicolon ( **;** )
- Exclamation mark ( **!** )
- Question mark ( **?** )
- Left Parentheses ( **(** )
- Right Parentheses ( **)** )
- Dash ( **-** )
- Return ( **\n** )

This dictionary will be used to tokenize the symbols and add the delimiter (space) around it.  This separates each symbols as its own word, making it easier for the neural network to predict the next word. Make sure you don't use a value that could be confused as a word; for example, instead of using the value "dash", try using something like "||dash||".

In [5]:
def token_lookup():
    """
    Generate a dict to turn punctuation into a token. Includes all from punctuation
    from string import punctuation;print(punctuation); and some more
    :return: Tokenized dictionary where the key is the punctuation and the value is the token
    """
    # TODO: Implement Function
    punct_lookup = {
        '!':'<EXCLAMATION>',
        '"':'<DOUBLE_QUOTE>',
        '#':'<HASH>',
        '$':'<DOLLAR>',
        '%':'<PERCENT>',
        '&':'<AND>',
        "'":'<APOSTROPHE>',
        '(':'<OPEN_BRACKET>',
        ')':'<CLOSE_BRACKET>',
        '*':'<ASTERISK>',
        '+':'<PLUS>',
        ',':'<COMMA>',
        '-':'<HYPHEN>',
        '.':'<STOP>',
        '/':'<FWD_SLASH>',
        ';':'<SEMI_COLON>',
        '<':'<OPEN_ANGLE>',
        '=':'<EQUALS>',
        '>':'<CLOSE_ANGLE>',
        '?':'<QUESTION_MARK>',
        '@':'<AT>',
        '[':'<OPEN_SQUARE>',
        '\\':'<BACKSLASH>',
        ']':'<CLOSE_SQUARE>',
        '^':'<HAT>',
        '_':'<UNDERSCORE>',
        '`':'<SINGLE_QUOTE>',
        '{':'<OPEN_CURLY>',
        '|':'<PIPE>',
        '}':'<CLOSE_CURLY>',
        '~':'<TILDE>',
        '\n':'<NEW_LINE>',
        '\t':'<TAB>',
        '\r':'<CARRIAGE_RETURN>',
    }

    subset = ['-', '?', '.', '\n', '(', ';', '!', ',', ')', '"']

        
    return {key: punct_lookup[key] for key in subset}

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_tokenize(token_lookup)

Tests Passed


## Pre-process all the data and save it

Running the code cell below will pre-process all the data and save it to file. You're encouraged to look at the code for `preprocess_and_save_data` in the `helpers.py` file to see what it's doing in detail, but you do not need to change this code.

In [6]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# pre-process training data
helper.preprocess_and_save_data(data_dir, token_lookup, create_lookup_tables)

In [7]:
# This is extract from first step of pre-process above....
#... just to check what its doing before we convert to ints and
# save as binary file

text = helper.load_data(data_dir)

# Ignore notice, since we don't use it for analysing the data
text = text[81:]

token_dict = token_lookup()
for key, token in token_dict.items():
    text = text.replace(key, ' {} '.format(token))

text = text.lower()
text = text.split()

print(text[100:500])

['where', 'did', 'he', 'go', '<question_mark>', 'he', 'didnt', 'tell', 'me', 'where', 'he', 'was', 'going', '<stop>', 'he', 'must', 'have', 'gone', 'out', '<stop>', 'you', 'wanna', 'go', 'out', 'you', 'get', 'ready', '<comma>', 'you', 'pick', 'out', 'the', 'clothes', '<comma>', 'right', '<question_mark>', 'you', 'take', 'the', 'shower', '<comma>', 'you', 'get', 'all', 'ready', '<comma>', 'get', 'the', 'cash', '<comma>', 'get', 'your', 'friends', '<comma>', 'the', 'car', '<comma>', 'the', 'spot', '<comma>', 'the', 'reservation', '<stop>', '<stop>', '<stop>', 'then', 'youre', 'standing', 'around', '<comma>', 'what', 'do', 'you', 'do', '<question_mark>', 'you', 'go', 'we', 'gotta', 'be', 'getting', 'back', '<stop>', 'once', 'youre', 'out', '<comma>', 'you', 'wanna', 'get', 'back', '<exclamation>', 'you', 'wanna', 'go', 'to', 'sleep', '<comma>', 'you', 'wanna', 'get', 'up', '<comma>', 'you', 'wanna', 'go', 'out', 'again', 'tomorrow', '<comma>', 'right', '<question_mark>', 'where', 'ever', 

# Check Point
This is your first checkpoint. If you ever decide to come back to this notebook or have to restart the notebook, you can start from here. The preprocessed data has been saved to disk.

In [8]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import helper
import problem_unittests as tests

int_text, vocab_to_int, int_to_vocab, token_dict = helper.load_preprocess()

In [9]:
# Make sure special padding token is there
print(vocab_to_int['<PAD>'])
print(int_to_vocab[0])

11845
fooled


## Build the Neural Network
In this section, you'll build the components necessary to build an RNN by implementing the RNN Module and forward and backpropagation functions.

### Check Access to GPU

In [10]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import torch

# Check for a GPU
train_on_gpu = torch.cuda.is_available()
if not train_on_gpu:
    print('No GPU found. Please use a GPU to train your neural network.')

## Input
Let's start with the preprocessed input data. We'll use [TensorDataset](http://pytorch.org/docs/master/data.html#torch.utils.data.TensorDataset) to provide a known format to our dataset; in combination with [DataLoader](http://pytorch.org/docs/master/data.html#torch.utils.data.DataLoader), it will handle batching, shuffling, and other dataset iteration functions.

You can create data with TensorDataset by passing in feature and target tensors. Then create a DataLoader as usual.
```
data = TensorDataset(feature_tensors, target_tensors)
data_loader = torch.utils.data.DataLoader(data, 
                                          batch_size=batch_size)
```

### Batching
Implement the `batch_data` function to batch `words` data into chunks of size `batch_size` using the `TensorDataset` and `DataLoader` classes.

>You can batch words using the DataLoader, but it will be up to you to create `feature_tensors` and `target_tensors` of the correct size and content for a given `sequence_length`.

For example, say we have these as input:
```
words = [1, 2, 3, 4, 5, 6, 7]
sequence_length = 4
```

Your first `feature_tensor` should contain the values:
```
[1, 2, 3, 4]
```
And the corresponding `target_tensor` should just be the next "word"/tokenized word value:
```
5
```
This should continue with the second `feature_tensor`, `target_tensor` being:
```
[2, 3, 4, 5]  # features
6             # target
```


In [11]:
# Trial run for dataloading methodology:

test = list(range(20))
print(len(test))
print(test,'\n')


test_feat = [test[i:i+5] for i in range(len(test)-5)]
print(test_feat,'\n')
print(torch.Tensor(test_feat).shape)

test_targ = [test[i+5] for i in range(len(test)-5)]
print(test_targ,'\n')
print(torch.Tensor(test_targ).shape)

20
[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19] 

[[0, 1, 2, 3, 4], [1, 2, 3, 4, 5], [2, 3, 4, 5, 6], [3, 4, 5, 6, 7], [4, 5, 6, 7, 8], [5, 6, 7, 8, 9], [6, 7, 8, 9, 10], [7, 8, 9, 10, 11], [8, 9, 10, 11, 12], [9, 10, 11, 12, 13], [10, 11, 12, 13, 14], [11, 12, 13, 14, 15], [12, 13, 14, 15, 16], [13, 14, 15, 16, 17], [14, 15, 16, 17, 18]] 

torch.Size([15, 5])
[5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19] 

torch.Size([15])


In [25]:
from torch.utils.data import TensorDataset, DataLoader

def batch_data(words, sequence_length, batch_size):
    """
    Batch the neural network data using DataLoader
    :param words: The word ids of the TV scripts
    :param sequence_length: The sequence length of each batch
    :param batch_size: The size of each batch; the number of sequences in a batch
    :return: DataLoader with batched data
    """
    if torch.cuda.is_available():
        num_workers = 4
    else:
        num_workers = 0

    # Truncate to make only full batches
    n_full_seq = len(words) // sequence_length

    # handle edge case
    # make sure that there is one-word left to
    # attach corresponding label to feature
    if n_full_seq * sequence_length + 1 > len(words):
        n_full_seq -= 1

    # rolling window through words
    # append on lists is faster than on np arrays so better to do on list first
    features = [words[i:i + sequence_length] for i in range((n_full_seq - 1) * sequence_length)]
    targets = [words[i + sequence_length] for i in range((n_full_seq - 1) * sequence_length)]

    dataset = TensorDataset(torch.LongTensor(features), torch.LongTensor(targets))
    loader = DataLoader(dataset, batch_size=batch_size, shuffle=True, num_workers=num_workers)

    return loader


### Test your dataloader 

You'll have to modify this code to test a batching function, but it should look fairly similar.

Below, we're generating some test text data and defining a dataloader using the function you defined, above. Then, we are getting some sample batch of inputs `sample_x` and targets `sample_y` from our dataloader.

Your code should return something like the following (likely in a different order, if you shuffled your data):

```
torch.Size([10, 5])
tensor([[ 28,  29,  30,  31,  32],
        [ 21,  22,  23,  24,  25],
        [ 17,  18,  19,  20,  21],
        [ 34,  35,  36,  37,  38],
        [ 11,  12,  13,  14,  15],
        [ 23,  24,  25,  26,  27],
        [  6,   7,   8,   9,  10],
        [ 38,  39,  40,  41,  42],
        [ 25,  26,  27,  28,  29],
        [  7,   8,   9,  10,  11]])

torch.Size([10])
tensor([ 33,  26,  22,  39,  16,  28,  11,  43,  30,  12])
```

### Sizes
Your sample_x should be of size `(batch_size, sequence_length)` or (10, 5) in this case and sample_y should just have one dimension: batch_size (10). 

### Values

You should also notice that the targets, sample_y, are the *next* value in the ordered test_text data. So, for an input sequence `[ 28,  29,  30,  31,  32]` that ends with the value `32`, the corresponding output should be `33`.

In [26]:
# test dataloader

test_text = range(51)
t_loader = batch_data(test_text, sequence_length=5, batch_size=10)
print(len(t_loader))
data_iter = iter(t_loader)
sample_x, sample_y = data_iter.next()

print(sample_x.shape)
print(sample_x)
print()
print(sample_y.shape)
print(sample_y)

5
torch.Size([10, 5])
tensor([[24, 25, 26, 27, 28],
        [ 5,  6,  7,  8,  9],
        [39, 40, 41, 42, 43],
        [12, 13, 14, 15, 16],
        [16, 17, 18, 19, 20],
        [19, 20, 21, 22, 23],
        [ 9, 10, 11, 12, 13],
        [ 2,  3,  4,  5,  6],
        [17, 18, 19, 20, 21],
        [21, 22, 23, 24, 25]])

torch.Size([10])
tensor([29, 10, 44, 17, 21, 24, 14,  7, 22, 26])


I bought this cell fwd because I like to sample some text as I go

In [27]:
"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
import torch.nn.functional as F

def generate(rnn, prime_id, int_to_vocab, token_dict, pad_value, predict_len=100):
    """
    Generate text using the neural network
    :param prime_id: The word id to start the first prediction
    :param int_to_vocab: Dict of word id keys to word values
    :param token_dict: Dict of puncuation tokens keys to puncuation values
    :param pad_value: The value used to pad a sequence
    :param predict_len: The length of text to generate
    :return: The generated text
    """
    rnn.eval()

    # create a sequence (batch_size=1) with the prime_id
    current_seq = np.full((1, sequence_length), pad_value)
    current_seq[-1][-1] = prime_id
    predicted = [int_to_vocab[prime_id]]

    for _ in range(predict_len):
        if train_on_gpu:
            current_seq = torch.LongTensor(current_seq).cuda()
        else:
            current_seq = torch.LongTensor(current_seq)

        # initialize the hidden state
        hidden = rnn.init_hidden(current_seq.size(0))

        # get the output of the rnn
        output, _ = rnn(current_seq, hidden)

        # get the next word probabilities
        p = F.softmax(output, dim=1).data
        if(train_on_gpu):
            p = p.cpu() # move to cpu

        # use top_k sampling to get the index of the next word
        top_k = 5
        p, top_i = p.topk(top_k)
        top_i = top_i.numpy().squeeze()

        # select the likely next word index with some element of randomness
        p = p.numpy().squeeze()
        word_i = np.random.choice(top_i, p=p/p.sum())

        # retrieve that word from the dictionary
        word = int_to_vocab[word_i]
        predicted.append(word)

        if(train_on_gpu):
            current_seq = current_seq.cpu() # move to cpu
        # the generated word becomes the next "current sequence" and the cycle can continue
        if train_on_gpu:
            current_seq = current_seq.cpu()
        current_seq = np.roll(current_seq, -1, 1)
        current_seq[-1][-1] = word_i

    gen_sentences = ' '.join(predicted)

    # Replace punctuation tokens
    for key, token in token_dict.items():
        ending = ' ' if key in ['\n', '(', '"'] else ''
        gen_sentences = gen_sentences.replace(' ' + token.lower(), key)
    gen_sentences = gen_sentences.replace('\n ', '\n')
    gen_sentences = gen_sentences.replace('( ', '(')

    # return all the sentences
    return gen_sentences


---
## Build the Neural Network
Implement an RNN using PyTorch's [Module class](http://pytorch.org/docs/master/nn.html#torch.nn.Module). You may choose to use a GRU or an LSTM. To complete the RNN, you'll have to implement the following functions for the class:
 - `__init__` - The initialize function. 
 - `init_hidden` - The initialization function for an LSTM/GRU hidden state
 - `forward` - Forward propagation function.
 
The initialize function should create the layers of the neural network and save them to the class. The forward propagation function will use these layers to run forward propagation and generate an output and a hidden state.

**The output of this model should be the *last* batch of word scores** after a complete sequence has been processed. That is, for each input sequence of words, we only want to output the word scores for a single, most likely, next word.

### Hints

1. Make sure to stack the outputs of the lstm to pass to your fully-connected layer, you can do this with `lstm_output = lstm_output.contiguous().view(-1, self.hidden_dim)`
2. You can get the last batch of word scores by shaping the output of the final, fully-connected layer like so:

```
# reshape into (batch_size, seq_length, output_size)
output = output.view(batch_size, -1, self.output_size)
# get last batch
out = output[:, -1]
```

In [28]:
import torch.nn as nn

class RNN(nn.Module):

    def __init__(self, vocab_size, output_size, embedding_dim, hidden_dim, n_layers, dropout=0.5):
        """
        Initialize the PyTorch RNN Module
        :param vocab_size: The number of input dimensions of the neural network (the size of the vocabulary)
        :param output_size: The number of output dimensions of the neural network
        :param embedding_dim: The size of embeddings, should you choose to use them
        :param hidden_dim: The size of the hidden layer outputs
        :param dropout: dropout to add in between LSTM/GRU layers
        """
        super(RNN, self).__init__()

        # set class variables
        # todo Problematic if save whole rnn and load on different device
        self.train_on_gpu = torch.cuda.is_available()

        self.output_size = output_size
        self.n_layers = n_layers
        self.hidden_dim = hidden_dim

        # layers
        self.embedding = nn.Embedding(vocab_size, embedding_dim)
        self.lstm = nn.LSTM(embedding_dim, hidden_dim, n_layers,
                            dropout=dropout, batch_first=True)

        self.dropout = nn.Dropout(dropout)

        self.fc = nn.Linear(hidden_dim, output_size)

    def forward(self, nn_input, hidden):
        """
        Forward propagation of the neural network
        :param nn_input: The input to the neural network
        :param hidden: The hidden state
        :return: Two Tensors, the output of the neural network and the latest hidden state
        """

        # embeddings and lstm_out
        nn_input = nn_input.long()
        embeds = self.embedding(nn_input)
        lstm_out, hidden = self.lstm(embeds, hidden)

        # stack up lstm outputs
        lstm_out = lstm_out.contiguous().view(-1, self.hidden_dim)

        # dropout and fully-connected layer
        out = self.dropout(lstm_out)
        out = self.fc(out)

        # reshape into (batch_size, seq_length, output_size)
        batch_size = nn_input.size(0)
        out = out.view(batch_size, -1, self.output_size)
        out = out[:, -1]  # get last word in sequence

        return out, hidden

    def init_hidden(self, batch_size):
        '''
        Initialize the hidden state of an LSTM/GRU
        :param batch_size: The batch_size of the hidden state
        :return: hidden state of dims (n_layers, batch_size, hidden_dim)
        '''
        # Create two new tensors with sizes n_layers x batch_size x hidden_dim,
        # initialized to zero, for hidden state and cell state of LSTM
        weight = next(self.parameters()).data

        if (self.train_on_gpu):
            hidden = (weight.new(self.n_layers, batch_size, self.hidden_dim).zero_().cuda(),
                      weight.new(self.n_layers, batch_size, self.hidden_dim).zero_().cuda())
        else:
            hidden = (weight.new(self.n_layers, batch_size, self.hidden_dim).zero_(),
                      weight.new(self.n_layers, batch_size, self.hidden_dim).zero_())

        # initialize hidden state with zero weights, and move to GPU if available

        return hidden

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_rnn(RNN, train_on_gpu)

Tests Passed


### Define forward and backpropagation

Use the RNN class you implemented to apply forward and back propagation. This function will be called, iteratively, in the training loop as follows:
```
loss = forward_back_prop(decoder, decoder_optimizer, criterion, inp, target)
```

And it should return the average loss over a batch and the hidden state returned by a call to `RNN(inp, hidden)`. Recall that you can get this loss by computing it, as usual, and calling `loss.item()`.

**If a GPU is available, you should move your data to that GPU device, here.**

In [29]:
def forward_back_prop(rnn, optimizer, criterion, inp, target, hidden):
    """
    Forward and backward propagation on the neural network
    :param rnn: The PyTorch Module that holds the neural network
    :param optimizer: The PyTorch optimizer for the neural network
    :param criterion: The PyTorch loss function
    :param inp: A batch of input to the neural network
    :param target: The target output for the batch of input
    :return: The loss and the latest hidden state Tensor
    """

    # move data to GPU, if available
    if torch.cuda.is_available():
        inp, target = inp.cuda(), target.cuda()

    ## perform backpropagation and optimization
    # Creating new variables for the hidden state, otherwise
    # we'd backprop through the entire training history
    h = tuple([each.data for each in hidden])

    # zero accumulated gradients
    rnn.zero_grad()

    # get the output from the fwd pass
    output, h = rnn.forward(inp, h)

    # calculate the loss and perform backprop
    # output is scores for each word in vocab, target is index position of correct word in vocab
    loss = criterion(output, target)
    loss.backward()

    #`clip_grad_norm` helps prevent the exploding gradient problem in RNNs / LSTMs.
    nn.utils.clip_grad_norm_(rnn.parameters(), 5.)
    optimizer.step()

    # return the loss over a batch and the hidden state produced by our model
    return loss.item(), hidden

# Note that these tests aren't completely extensive.
# they are here to act as general checks on the expected outputs of your functions
"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_forward_back_prop(RNN, forward_back_prop, train_on_gpu)

Tests Passed


## Neural Network Training

With the structure of the network complete and data ready to be fed in the neural network, it's time to train it.

### Train Loop

The training loop is implemented for you in the `train_decoder` function. This function will train the network over all the batches for the number of epochs given. The model progress will be shown every number of batches. This number is set with the `show_every_n_batches` parameter. You'll set this parameter along with other parameters in the next section.

In [17]:
import time
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""

def train_rnn(rnn, batch_size, optimizer, criterion, n_epochs, show_every_n_batches=100):
    batch_losses = []

    rnn.train()
    with open('log_'+'{}'.format(time.time())+'.txt', 'w') as log_file: #
        print("Training for %d epoch(s)..." % n_epochs)
        for epoch_i in range(1, n_epochs + 1):
            start = time.time()
            begin_loss = 0

            # initialize hidden state
            hidden = rnn.init_hidden(batch_size)

            for batch_i, (inputs, labels) in enumerate(train_loader, 1):

                # make sure you iterate over completely full batches, only
                n_batches = len(train_loader.dataset) // batch_size
                if (batch_i > n_batches):
                    break

                # forward, back prop
                loss, hidden = forward_back_prop(rnn, optimizer, criterion, inputs, labels, hidden)
                # record loss
                batch_losses.append(loss)

                # running stats
                elapsed_time = float(time.time() - start)
                speed = batch_size * batch_i / elapsed_time if elapsed_time > 0 else 0
                log_txt = "E{} B{} {:.0f}% \tTraining Loss: {:.6f} --> {:.6f} | avg {:.6f} \t#/s: {:.1f}".format(
                    epoch_i, batch_i, 100. * batch_i / n_batches, batch_losses[0], batch_losses[-1], np.average(batch_losses), speed)
                print(log_txt, end='\r')
                # intermittent save checkpoint & sample
                if (batch_i+(epoch_i-1)*n_batches) % show_every_n_batches == 0:
                    log_file.write(log_txt+'\n')
                    torch.save(rnn.state_dict(), 'trained_rnn_sd_'+format(np.average(batch_losses),'.6f').replace('.','_'))
                    generated_script = generate(rnn, vocab_to_int['jerry:'], int_to_vocab, token_dict,
                                                vocab_to_int[helper.SPECIAL_WORDS['PADDING']], 100)

                    log_txt = '\n' + generated_script.replace('\n', ' | ') + '\n\n'
                    log_file.write(log_txt)
                    print(log_txt)
                    rnn.train()

            # epoch stats
            print('Epoch: {:>4}/{:<4} Loss (beg avg end): {} {} {}\n'.format(
                epoch_i, n_epochs, batch_losses[0], np.average(batch_losses), batch_losses[-1]))
            batch_losses = []

    return rnn

### Hyperparameters

Set and train the neural network with the following parameters:
- Set `sequence_length` to the length of a sequence.
- Set `batch_size` to the batch size.
- Set `num_epochs` to the number of epochs to train for.
- Set `learning_rate` to the learning rate for an Adam optimizer.
- Set `vocab_size` to the number of unique tokens in our vocabulary.
- Set `output_size` to the desired size of the output.
- Set `embedding_dim` to the embedding dimension; smaller than the vocab_size.
- Set `hidden_dim` to the hidden dimension of your RNN.
- Set `n_layers` to the number of layers/cells in your RNN.
- Set `show_every_n_batches` to the number of batches at which the neural network should print progress.

If the network isn't getting the desired results, tweak these parameters and/or the layers in the `RNN` class.

In [37]:
resume = False

train_on_gpu = torch.cuda.is_available()

int_text, vocab_to_int, int_to_vocab, token_dict = helper.load_preprocess()
# Data params
# Sequence Length
sequence_length = 20  # of words in a sequence
# Batch Size
batch_size = 128

# data loader - do not change
train_loader = batch_data(int_text, sequence_length, batch_size)

In [38]:
# Training parameters
# Number of Epochs
num_epochs = 15
# Learning Rate
learning_rate = 0.001

# Model parameters
# Vocab size
vocab_size = len(vocab_to_int)
# Output size
output_size = vocab_size
# Embedding Dimension
embedding_dim = 256
# Hidden Dimension
hidden_dim = 256
# Number of RNN Layers
n_layers = 2

# Show stats for every n number of batches
# Now per x seconds elapsed
show_every_n_batches = 2500

### Train
In the next cell, you'll train the neural network on the pre-processed data.  If you have a hard time getting a good loss, you may consider changing your hyperparameters. In general, you may get better results with larger hidden and n_layer dimensions, but larger models take a longer time to train. 
> **You should aim for a loss less than 3.5.** 

You should also experiment with different sequence lengths, which determine the size of the long range dependencies that a model can learn.

In [20]:
import numpy as np

# create model and move to gpu if available
rnn = RNN(vocab_size, output_size, embedding_dim, hidden_dim, n_layers, dropout=0.25)

if resume:
    if train_on_gpu:
        rnn.load_state_dict(torch.load('trained_rnn_sd', map_location=torch.device('cuda')))
    else:
        rnn.load_state_dict(torch.load('trained_rnn_sd', map_location=torch.device('cpu')))
else:
    print('Starting with new model...')
if train_on_gpu:
    rnn.cuda()

generated_script = generate(rnn, vocab_to_int['jerry:'], int_to_vocab, token_dict,
                                vocab_to_int[helper.SPECIAL_WORDS['PADDING']], 100)
print('\n', generated_script.replace('\n', ' | '), '\n')

# defining loss and optimization functions for training
optimizer = torch.optim.Adam(rnn.parameters(), lr=learning_rate)
criterion = nn.CrossEntropyLoss()

# training the model
try:
    trained_rnn = train_rnn(rnn, batch_size, optimizer, criterion, num_epochs, show_every_n_batches)
except KeyboardInterrupt:
    torch.save(rnn.state_dict(), 'trained_rnn_sd')
    print('\nLatest Model Saved')

Starting with new model...

 jerry: granmont: innocence wants guest wants unless wants wants think unless smoothed wants wants wants beg york suspicious wants york wants proud dysfunctional lupis suspicious movers keyed foreign wants foreign penant pleased “saree” penant sanctum sanctum smoothed 'please brad wants squint bloated squint bloated afuera anguished scoop martins olive: sneezed foreign delving penant lifting penant guarantees craggly subtitle craggly subtitle fleesher's accomplished confrontation wants wants snoopy tuscany] 'where's craggly summertime penant extremely quicksand bell's tie in in waj handed 1990 extremely blades dolphin patience danny: saaage treats aha movers unphased cooling ehm unless speed yeay wants ties 'where's intercom] unless barber 

Training for 15 epoch(s)...
E1 B2500 36% 	Training Loss: 9.956375 --> 3.892861 | avg 4.811303 	#/s: 1727.4
jerry: heyyyyyy gladys: vivian: i don't know what i have to have to do that? i don't have to be. |  | jerry: i wa

Epoch:    6/15   Loss (beg avg end): 3.3012049198150635 3.613479332835624 3.3622639179229736

E7 B686 10% 	Training Loss: 3.519839 --> 3.865047 | avg 3.520395 	#/s: 1707.8
jerry: fleeing stumbled, and then we have to be so funny to be in the mood. |  | jerry: well, i was thinking about a very bad idea, but i'm going to get out of the car. |  | george:(smiling) no, no. i'm sorry. i don't have it. |  | elaine:(to kramer) what happened? |  | elaine: i got some kind of a good time, you can be ashamed. |  | kramer: yeah, yeah, i'm sure i could get a good


E7 B3186 46% 	Training Loss: 3.519839 --> 3.527619 | avg 3.538344 	#/s: 1710.2
jerry: holy boo busting frank: busting vivian 915 penis: mayo: hahaha!!!(fuck). |  | elaine:(yelling) i know you, i don't have any time.(jerry enters) |  | elaine:(yelling out) oh! |  | jerry:(to jerry and george) you know what? i'm gonna get back to the airport. |  | jerry: oh, i think i'm gonna get the car. i can't believe this... i was a little tired. |  | g

Epoch:   12/15   Loss (beg avg end): 3.2818121910095215 3.393848489014725 3.126473903656006

E13 B1372 20% 	Training Loss: 3.388216 --> 3.081261 | avg 3.305402 	#/s: 1637.3
jerry: goodness peterman: squinting liver squinting hed come in. ] |  | george:(to kramer) you know how you can get you a new place. |  | george: i can't go. |  | elaine:(entering, to elaine) oh, yeah.(to jerry) you know, it's not the point. |  | jerry:(to jerry) you know, i think i'm going to get to the airport for a while... |  | jerry:(to jerry) you know, the whole thing is,


E13 B3872 56% 	Training Loss: 3.388216 --> 3.101094 | avg 3.341112 	#/s: 1636.5
jerry: banned discuss the gosh grew. |  | elaine: oh, you know i don't want to see that. |  | kramer: oh, i can't. |  | jerry: oh, i'm not going to do that. i'm not going to see you. i don't want to. |  | jerry: well, i don't know if i was going to tell me that he didn't want it. i don't think i'm a good driver, but i don't know what you think. |  | george: i me

Process Process-61:
Process Process-62:
Process Process-63:
Process Process-64:
Traceback (most recent call last):
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/opt/conda/lib/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/opt/conda/lib/python3.6/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
Traceback (most recent call last):
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/site-packages/torch/utils/data/dataloader.py", line 52, in _worker_loop
    r = index_queue.get()
  File "/opt/conda/lib/python3.6/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
  File "/opt/conda/lib/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/opt/conda/lib/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run

  File "/opt/conda/lib/python3.6/multiprocessing/connection.py", line 379, in _recv
    chunk = read(handle, remaining)
KeyboardInterrupt
KeyboardInterrupt


RuntimeError: DataLoader worker (pid 51) exited unexpectedly with exit code 1.

### Question: How did you decide on your model hyperparameters? 
For example, did you try different sequence_lengths and find that one size made the model converge faster? What about your hidden_dim and n_layers; how did you decide on those?

**Answer:** (Write answer, here)
I started with Karpathy's blog post on RNNs which gave useful guidance for a starting point. He used 512 for hidden dims and 2-layers for small datasets, and 3-layers for larger ones. His advice is to match the number of features with the number of characters. In our case we're looking at about 3.5m characters. Karpathy used 100 seq length, but said you could double speed with minimal performance cost by going to 50.

Unfortunately, when starting with these larger hp's my network was converging very slowly. It looked like it would not get below 3.5 in a reasonable time. I halved the dims to 256, and reduced dropout from 50 to 25% to help convergence. Since this implementation is not monitoring validation error, I will risk overfitting for quicker convergence. Still, the network was taking a while. I trimmed seq length to 20 and this worked the best.

I realize in hindsight that I did not perform gradient clipping. This could explain the slow convergence in my larger models. Thankfull, I managed to get a decent network without. If I have spare GPU time left, i will retry the larger models with clipping enabled.

I'm also curious about trying to train the embeddings first with skip-grams / 

---
# Checkpoint

After running the above training cell, your model will be saved by name, `trained_rnn`, and if you save your notebook progress, **you can pause here and come back to this code at another time**. You can resume your progress by running the next cell, which will load in our word:id dictionaries _and_ load in your saved model by name!

In [44]:
import torch
import helper

_, vocab_to_int, int_to_vocab, token_dict = helper.load_preprocess()
#trained_rnn = helper.load_model('./save/trained_rnn')
trained_rnn = RNN(vocab_size, output_size, embedding_dim, hidden_dim, n_layers, dropout=0.25)

if train_on_gpu:
    trained_rnn.load_state_dict(torch.load('trained_rnn_sd', map_location=torch.device('cuda')))
else:
    trained_rnn.load_state_dict(torch.load('trained_rnn_sd', map_location=torch.device('cpu')))

## Generate TV Script
With the network trained and saved, you'll use it to generate a new, "fake" Seinfeld TV script in this section.

### Generate Text
To generate the text, the network needs to start with a single word and repeat its predictions until it reaches a set length. You'll be using the `generate` function to do this. It takes a word id to start with, `prime_id`, and generates a set length of text, `predict_len`. Also note that it uses topk sampling to introduce some randomness in choosing the most likely next word, given an output set of word scores!

### Generate a New Script
It's time to generate the text. Set `gen_length` to the length of TV script you want to generate and set `prime_word` to one of the following to start the prediction:
- "jerry"
- "elaine"
- "george"
- "kramer"

You can set the prime word to _any word_ in our dictionary, but it's best to start with a name for generating a TV script. (You can also start with any other names you find in the original text file!)

In [47]:
import numpy as np
# run the cell multiple times to get different results!
gen_length = 400 # modify the length to your preference
prime_word = 'jerry' # name for starting the script

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
pad_word = helper.SPECIAL_WORDS['PADDING']
generated_script = generate(trained_rnn, vocab_to_int[prime_word + ':'], int_to_vocab, token_dict, vocab_to_int[pad_word], gen_length)
print(generated_script)

jerry: on the phone)

george: what do you think of that?

jerry: i know, i don't know...

george:(interrupting) what?

jerry:(to george) i got it! i can't believe this, i'm sorry. i can't take it.

elaine: i don't know how you can do this.

jerry:(looking at his watch) oh, yeah. i know... i don't know.

kramer: i don't know.

jerry: well, i was just curious.

george: what is it?

jerry: i know. i know you have a big salad.

george: what about a woman?

jerry: i know.(to kramer) hey, what are they doing with that?

kramer: yeah, well, i just want to go to the bathroom.

kramer: hey.

elaine: oh, i'm sorry i can't!

george: you know what, you can't do that?

jerry: no, no.

elaine: well, you know, i was just thinking. you know, i just don't know if i was to get a little..

elaine: oh, you can't do it.

jerry: i can't believe you don't want to be able to get me to be a lot of work. you don't want to go to the bathroom.

george: i don't know.

george: you know, it's not a big problem.

jer

#### Save your favorite scripts

Once you have a script that you like (or find interesting), save it to a text file!

In [ ]:
# save script to a text file
f =  open("generated_script_3.txt","w")
f.write(generated_script)
f.close()

# The TV Script is Not Perfect
It's ok if the TV script doesn't make perfect sense. It should look like alternating lines of dialogue, here is one such example of a few generated lines.

### Example generated script

>jerry: what about me?
>
>jerry: i don't have to wait.
>
>kramer:(to the sales table)
>
>elaine:(to jerry) hey, look at this, i'm a good doctor.
>
>newman:(to elaine) you think i have no idea of this...
>
>elaine: oh, you better take the phone, and he was a little nervous.
>
>kramer:(to the phone) hey, hey, jerry, i don't want to be a little bit.(to kramer and jerry) you can't.
>
>jerry: oh, yeah. i don't even know, i know.
>
>jerry:(to the phone) oh, i know.
>
>kramer:(laughing) you know...(to jerry) you don't know.

You can see that there are multiple characters that say (somewhat) complete sentences, but it doesn't have to be perfect! It takes quite a while to get good results, and often, you'll have to use a smaller vocabulary (and discard uncommon words), or get more data.  The Seinfeld dataset is about 3.4 MB, which is big enough for our purposes; for script generation you'll want more than 1 MB of text, generally. 

# Submitting This Project
When submitting this project, make sure to run all the cells before saving the notebook. Save the notebook file as "dlnd_tv_script_generation.ipynb" and save another copy as an HTML file by clicking "File" -> "Download as.."->"html". Include the "helper.py" and "problem_unittests.py" files in your submission. Once you download these files, compress them into one zip file for submission.